File: packages/sdk-csharp/src/Primust/RecordInput.cs
namespace Primust;

/// <summary>
/// Input to Run.RecordAsync(). Provide check name, manifest, input bytes,
/// result, and optional details/output.
///
/// <code>
/// await run.RecordAsync(new RecordInput
/// {
///     Check = "coverage_verification",
///     ManifestId = "sha256:abc...",
///     Input = Encoding.UTF8.GetBytes(inputBinding),
///     CheckResult = CheckResult.Pass,
///     Details = new Dictionary&lt;string, object&gt; { ["claim_id"] = "CLM-001" },
///     Visibility = "opaque"
/// });
/// </code>
/// </summary>
public class RecordInput
{
    /// <summary>Governance check name (e.g., "coverage_verification").</summary>
    public required string Check { get; init; }

    /// <summary>Manifest identifier (SHA-256 hash or registered ID).</summary>
    public required string ManifestId { get; init; }

    /// <summary>
    /// Raw input bytes — committed locally via SHA-256/Poseidon2 before any transmission.
    /// The raw bytes NEVER leave your environment.
    /// </summary>
    public required byte[] Input { get; init; }

    /// <summary>Optional output bytes for verification (committed locally).</summary>
    public byte[]? Output { get; init; }

    /// <summary>Result of the governance check.</summary>
    public required CheckResult CheckResult { get; init; }

    /// <summary>Bounded metadata — key-value pairs. Must not contain sensitive data.</summary>
    public Dictionary<string, object>? Details { get; init; }

    /// <summary>Data visibility: "opaque" (default), "selective", or "transparent".</summary>
    public string Visibility { get; init; } = "opaque";
}

File: sdk-java/src/main/java/com/primust/sdk/RecordInput.java
package com.primust.sdk;

import java.util.Collections;
import java.util.Map;

/**
 * Input to Run.record() or Pipeline.record().
 * Use RecordInput.builder() to construct.
 *
 * <pre>
 * run.record(
 *     RecordInput.builder()
 *         .check("coverage_verification")
 *         .manifestId("sha256:abc...")
 *         .input(data.getBytes())
 *         .checkResult(CheckResult.PASS)
 *         .details(Map.of("claim_id", "CLM-001"))
 *         .visibility("opaque")
 *         .build()
 * );
 * </pre>
 */
public final class RecordInput {

    private final String check;
    private final String manifestId;
    private final byte[] input;
    private final byte[] output;
    private final CheckResult checkResult;
    private final Map<String, Object> details;
    private final String visibility;

    private RecordInput(Builder b) {
        if (b.check == null) throw new IllegalArgumentException("check is required");
        if (b.manifestId == null) throw new IllegalArgumentException("manifestId is required");
        if (b.input == null) throw new IllegalArgumentException("input is required");
        if (b.checkResult == null) throw new IllegalArgumentException("checkResult is required");
        this.check = b.check;
        this.manifestId = b.manifestId;
        this.input = b.input;
        this.output = b.output;
        this.checkResult = b.checkResult;
        this.details = b.details != null ? Collections.unmodifiableMap(b.details) : Collections.emptyMap();
        this.visibility = b.visibility != null ? b.visibility : "opaque";
    }

    public String check() { return check; }
    public String manifestId() { return manifestId; }
    public byte[] input() { return input; }
    public byte[] output() { return output; }
    public CheckResult checkResult() { return checkResult; }
    public Map<String, Object> details() { return details; }
    public String visibility() { return visibility; }

    public static Builder builder() { return new Builder(); }

    public static final class Builder {
        private String check;
        private String manifestId;
        private byte[] input;
        private byte[] output;
        private CheckResult checkResult;
        private Map<String, Object> details;
        private String visibility;

        public Builder check(String check) { this.check = check; return this; }
        public Builder manifestId(String manifestId) { this.manifestId = manifestId; return this; }
        public Builder input(byte[] input) { this.input = input; return this; }
        public Builder output(byte[] output) { this.output = output; return this; }
        public Builder checkResult(CheckResult checkResult) { this.checkResult = checkResult; return this; }
        public Builder checkResult(String checkResult) {
            this.checkResult = CheckResult.valueOf(checkResult.toUpperCase());
            return this;
        }
        public Builder details(Map<String, Object> details) { this.details = details; return this; }
        public Builder visibility(String visibility) { this.visibility = visibility; return this; }
        public Builder visibility(VisibilityMode visibility) { this.visibility = visibility.value(); return this; }

        public RecordInput build() { return new RecordInput(this); }
    }
}
